File: iot_driver_copilot/wheeltecros机器人/driver.py
import os
import yaml
import json
import threading
import time
from typing import Callable, Dict, Any
from flask import Flask, request, jsonify
import paho.mqtt.client as mqtt
from kubernetes import client as k8s_client, config as k8s_config

# CONSTANTS
CONFIG_PATH = "/etc/edgedevice/config/instructions"
INSTRUCTION_FILE = os.path.join(CONFIG_PATH, "instructions.yaml")
EDGEDEVICE_CRD_GROUP = "shifu.edgenesis.io"
EDGEDEVICE_CRD_VERSION = "v1alpha1"
EDGEDEVICE_CRD_PLURAL = "edgedevices"

# ENVIRONMENT VARIABLES
EDGEDEVICE_NAME = os.environ.get("EDGEDEVICE_NAME")
EDGEDEVICE_NAMESPACE = os.environ.get("EDGEDEVICE_NAMESPACE")
MQTT_BROKER_ADDRESS = os.environ.get("MQTT_BROKER_ADDRESS")
MQTT_KEEPALIVE = int(os.environ.get("MQTT_KEEPALIVE", "60"))
MQTT_USERNAME = os.environ.get("MQTT_USERNAME", "")
MQTT_PASSWORD = os.environ.get("MQTT_PASSWORD", "")
MQTT_CLIENT_ID = os.environ.get("MQTT_CLIENT_ID", "deviceshifu-car-mqtt")
MQTT_TLS = os.environ.get("MQTT_TLS", "false").lower() == "true"
MQTT_PREFIX = "deviceshifu-car-mqtt"

if not EDGEDEVICE_NAME or not EDGEDEVICE_NAMESPACE or not MQTT_BROKER_ADDRESS:
    raise Exception("Required environment variables: EDGEDEVICE_NAME, EDGEDEVICE_NAMESPACE, MQTT_BROKER_ADDRESS")

# K8s API Initialization
k8s_config.load_incluster_config()
custom_api = k8s_client.CustomObjectsApi()

def update_edgedevice_phase(phase: str):
    body = {"status": {"edgeDevicePhase": phase}}
    try:
        custom_api.patch_namespaced_custom_object_status(
            group=EDGEDEVICE_CRD_GROUP,
            version=EDGEDEVICE_CRD_VERSION,
            namespace=EDGEDEVICE_NAMESPACE,
            plural=EDGEDEVICE_CRD_PLURAL,
            name=EDGEDEVICE_NAME,
            body=body
        )
    except Exception:
        pass  # Optionally log error

# Load EdgeDevice Spec for device address
def fetch_device_address():
    try:
        ed = custom_api.get_namespaced_custom_object(
            EDGEDEVICE_CRD_GROUP,
            EDGEDEVICE_CRD_VERSION,
            EDGEDEVICE_NAMESPACE,
            EDGEDEVICE_CRD_PLURAL,
            EDGEDEVICE_NAME
        )
        return ed.get("spec", {}).get("address")
    except Exception:
        return None

# Load instruction settings
def load_instruction_settings():
    try:
        with open(INSTRUCTION_FILE, "r") as f:
            return yaml.safe_load(f)
    except Exception:
        return {}

instruction_settings = load_instruction_settings()

# MQTT Client Setup
mqtt_client = mqtt.Client(MQTT_CLIENT_ID, clean_session=True)
if MQTT_USERNAME:
    mqtt_client.username_pw_set(MQTT_USERNAME, MQTT_PASSWORD)
if MQTT_TLS:
    mqtt_client.tls_set()

# MQTT Event Handlers
mqtt_connected = threading.Event()
mqtt_failed = threading.Event()
mqtt_subscriptions = {}

def on_connect(client, userdata, flags, rc):
    if rc == 0:
        mqtt_connected.set()
        mqtt_failed.clear()
        update_edgedevice_phase("Running")
        for topic, callback in mqtt_subscriptions.items():
            qos = callback["qos"]
            client.subscribe(topic, qos)
    else:
        mqtt_connected.clear()
        mqtt_failed.set()
        update_edgedevice_phase("Failed")

def on_disconnect(client, userdata, rc):
    mqtt_connected.clear()
    if rc != 0:
        mqtt_failed.set()
        update_edgedevice_phase("Failed")
    else:
        update_edgedevice_phase("Pending")

def on_message(client, userdata, msg):
    topic = msg.topic
    if topic in mqtt_subscriptions:
        callback = mqtt_subscriptions[topic]["callback"]
        try:
            payload = msg.payload.decode("utf-8")
            data = json.loads(payload)
        except Exception:
            data = msg.payload
        callback(data)

mqtt_client.on_connect = on_connect
mqtt_client.on_disconnect = on_disconnect
mqtt_client.on_message = on_message

# EdgeDevice Phase Management Thread
def phase_monitor():
    while True:
        if not mqtt_connected.is_set():
            update_edgedevice_phase("Pending")
        time.sleep(10)

monitor_thread = threading.Thread(target=phase_monitor, daemon=True)
monitor_thread.start()

# MQTT Connect
def mqtt_connect():
    host, port = MQTT_BROKER_ADDRESS.split(":")
    mqtt_client.connect(host, int(port), MQTT_KEEPALIVE)
    mqtt_client.loop_start()

def mqtt_disconnect():
    mqtt_client.loop_stop()
    mqtt_client.disconnect()

# API Implementation
app = Flask(__name__)
received_data = {
    MQTT_PREFIX + "/odom": None,
    MQTT_PREFIX + "/scan": None,
    MQTT_PREFIX + "/joint_states": None,
    MQTT_PREFIX + "/point_cloud": None,
    MQTT_PREFIX + "/status": None,
}

def make_pub_api(endpoint: str, qos: int):
    @app.route(f"/{endpoint}", methods=["POST"])
    def handler():
        payload = request.get_json(force=True, silent=True) or {}
        topic = f"{MQTT_PREFIX}/{endpoint.split('/')[-1]}"
        settings = instruction_settings.get(endpoint, {}).get("protocolPropertyList", {})
        pub_qos = int(settings.get("qos", qos))
        ret = mqtt_client.publish(topic, json.dumps(payload), qos=pub_qos)
        if ret.rc == mqtt.MQTT_ERR_SUCCESS:
            return jsonify({"result": "published", "topic": topic}), 200
        else:
            return jsonify({"result": "publish_failed", "topic": topic}), 500
    handler.__name__ = f"api_{endpoint.replace('/', '_')}_pub"
    return handler

def make_sub_api(endpoint: str):
    @app.route(f"/{endpoint}", methods=["GET"])
    def handler():
        topic = f"{MQTT_PREFIX}/{endpoint.split('/')[-1]}"
        data = received_data.get(topic)
        if data is not None:
            return jsonify(data), 200
        else:
            return jsonify({"result": "no_data"}), 404
    handler.__name__ = f"api_{endpoint.replace('/', '_')}_sub"
    return handler

# MQTT Subscription Callback Registration
def sub_callback_factory(topic):
    def callback(data):
        received_data[topic] = data
    return callback

# API Endpoints Definition
pub_apis = [
    {"endpoint": "move/forward", "qos": 1},
    {"endpoint": "move/backward", "qos": 1},
    {"endpoint": "turn/left", "qos": 1},
    {"endpoint": "turn/right", "qos": 1},
    {"endpoint": "stop", "qos": 1},
]
sub_apis = [
    {"endpoint": "odom", "qos": 1},
    {"endpoint": "scan", "qos": 1},
    {"endpoint": "joint_states", "qos": 1},
    {"endpoint": "point_cloud", "qos": 1},
    {"endpoint": "status", "qos": 1},
]

# Register Publish APIs
for api in pub_apis:
    make_pub_api(api["endpoint"], api["qos"])

# Register Subscribe APIs and MQTT Subscriptions
for api in sub_apis:
    topic = f"{MQTT_PREFIX}/{api['endpoint']}"
    received_data[topic] = None
    mqtt_subscriptions[topic] = {
        "callback": sub_callback_factory(topic),
        "qos": api["qos"]
    }
    make_sub_api(api["endpoint"])

# MQTT Subscribe on Connect
def subscribe_topics():
    for topic, sub in mqtt_subscriptions.items():
        mqtt_client.subscribe(topic, sub["qos"])

# Main Run
if __name__ == "__main__":
    update_edgedevice_phase("Pending")
    device_address = fetch_device_address()
    if device_address is None:
        update_edgedevice_phase("Unknown")
    try:
        mqtt_connect()
        if mqtt_connected.wait(timeout=10):
            update_edgedevice_phase("Running")
        else:
            update_edgedevice_phase("Failed")
    except Exception:
        update_edgedevice_phase("Failed")

    subscribe_topics()
    app.run(host="0.0.0.0", port=int(os.environ.get("HTTP_PORT", "8080")), threaded=True)

File: iot_driver_copilot/wheeltecros机器人/driver.cpp
```cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <cstdlib>
#include <thread>
#include <mutex>
#include <atomic>
#include <chrono>
#include <csignal>
#include <condition_variable>

// YAML Parsing
#include <yaml-cpp/yaml.h>

// MQTT
#include "mqtt/async_client.h"

// Kubernetes In-Cluster API
#include <curl/curl.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

// DeviceShifu Constants
const std::string EDGEDEVICE_CRD_GROUP = "shifu.edgenesis.io";
const std::string EDGEDEVICE_CRD_VERSION = "v1alpha1";
const std::string EDGEDEVICE_CRD_PLURAL = "edgedevices";
const std::string EDGEDEVICE_STATUS_PHASE_PATH = "/status/edgeDevicePhase";

// DeviceShifu Phases
enum class EdgeDevicePhase { Pending, Running, Failed, Unknown };

std::string PhaseToString(EdgeDevicePhase phase) {
    switch (phase) {
        case EdgeDevicePhase::Pending: return "Pending";
        case EdgeDevicePhase::Running: return "Running";
        case EdgeDevicePhase::Failed: return "Failed";
        case EdgeDevicePhase::Unknown: return "Unknown";
    }
    return "Unknown";
}

// ========== Kubernetes In-Cluster API ==========

std::string GetServiceAccountToken() {
    std::ifstream token_file("/var/run/secrets/kubernetes.io/serviceaccount/token");
    if (!token_file) return "";
    std::stringstream buffer;
    buffer << token_file.rdbuf();
    return buffer.str();
}

std::string GetK8sCAPath() {
    return "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt";
}

std::string GetK8sAPIServer() {
    const char* host = std::getenv("KUBERNETES_SERVICE_HOST");
    const char* port = std::getenv("KUBERNETES_SERVICE_PORT");
    if (!host || !port) return "";
    return "https://" + std::string(host) + ":" + std::string(port);
}

class K8sClient {
public:
    K8sClient() {
        ca_cert = GetK8sCAPath();
        base_url = GetK8sAPIServer();
        token = GetServiceAccountToken();
    }

    // Patch status.edgeDevicePhase
    bool PatchEdgeDevicePhase(const std::string& namespace_,
                              const std::string& name,
                              EdgeDevicePhase phase) {
        std::string url = base_url + "/apis/" + EDGEDEVICE_CRD_GROUP + "/" +
                          EDGEDEVICE_CRD_VERSION + "/namespaces/" + namespace_ +
                          "/" + EDGEDEVICE_CRD_PLURAL + "/" + name + "/status";
        std::string patch = R"({"status":{"edgeDevicePhase":")" + PhaseToString(phase) + R"("}})";
        struct curl_slist *headers = NULL;
        headers = curl_slist_append(headers, "Content-Type: application/merge-patch+json");
        headers = curl_slist_append(headers, ("Authorization: Bearer " + token).c_str());

        CURL *curl = curl_easy_init();
        if (!curl) return false;
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_SSLCERTTYPE, "PEM");
        curl_easy_setopt(curl, CURLOPT_CAINFO, ca_cert.c_str());
        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1L);
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PATCH");
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, patch.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, patch.size());
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);

        CURLcode res = curl_easy_perform(curl);
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);

        return res == CURLE_OK;
    }

    // Get EdgeDevice .spec.address
    std::string GetEdgeDeviceAddress(const std::string& namespace_, const std::string& name) {
        std::string url = base_url + "/apis/" + EDGEDEVICE_CRD_GROUP + "/" +
                          EDGEDEVICE_CRD_VERSION + "/namespaces/" + namespace_ +
                          "/" + EDGEDEVICE_CRD_PLURAL + "/" + name;
        struct curl_slist *headers = NULL;
        headers = curl_slist_append(headers, ("Authorization: Bearer " + token).c_str());

        CURL *curl = curl_easy_init();
        if (!curl) return "";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_SSLCERTTYPE, "PEM");
        curl_easy_setopt(curl, CURLOPT_CAINFO, ca_cert.c_str());
        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1L);
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);

        std::string response;
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, [](char* ptr, size_t size, size_t nmemb, void* userdata) -> size_t {
            std::string* str = reinterpret_cast<std::string*>(userdata);
            size_t total = size * nmemb;
            str->append(ptr, total);
            return total;
        });
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);

        CURLcode res = curl_easy_perform(curl);
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);

        if (res != CURLE_OK) return "";

        try {
            auto j = json::parse(response);
            if (j.contains("spec") && j["spec"].contains("address")) {
                return j["spec"]["address"].get<std::string>();
            }
        } catch (...) {}
        return "";
    }

private:
    std::string ca_cert;
    std::string base_url;
    std::string token;
};

// ========== Configuration Loader ==========

struct APISettings {
    std::map<std::string, std::string> protocolPropertyList;
};

class InstructionConfig {
public:
    bool Load(const std::string& path) {
        try {
            YAML::Node root = YAML::LoadFile(path);
            for (auto it = root.begin(); it != root.end(); ++it) {
                std::string api = it->first.as<std::string>();
                APISettings settings;
                if (it->second["protocolPropertyList"]) {
                    for (auto p = it->second["protocolPropertyList"].begin();
                         p != it->second["protocolPropertyList"].end(); ++p) {
                        settings.protocolPropertyList[p->first.as<std::string>()] =
                            p->second.as<std::string>();
                    }
                }
                api_settings[api] = settings;
            }
            return true;
        } catch (...) {
            return false;
        }
    }

    const APISettings* GetSettings(const std::string& api) const {
        auto it = api_settings.find(api);
        if (it != api_settings.end()) return &(it->second);
        return nullptr;
    }

private:
    std::map<std::string, APISettings> api_settings;
};

// ========== MQTT Client Wrapper ==========

class MQTTClientWrapper {
public:
    MQTTClientWrapper(const std::string& broker, const std::string& client_id)
    : cli(broker, client_id), is_connected(false) {}

    bool Connect() {
        mqtt::connect_options connOpts;
        try {
            mqtt::token_ptr conntok = cli.connect(connOpts);
            conntok->wait();
            is_connected = true;
            return true;
        } catch (...) {
            is_connected = false;
            return false;
        }
    }

    void Disconnect() {
        try {
            if (is_connected) {
                cli.disconnect()->wait();
            }
            is_connected = false;
        } catch (...) {}
    }

    bool IsConnected() const {
        return is_connected;
    }

    // Subscribe to a topic and provide a callback
    bool Subscribe(const std::string& topic, int qos,
                   std::function<void(const std::string&)> cb) {
        if (!is_connected) return false;
        try {
            cli.set_callback([cb](mqtt::const_message_ptr msg) {
                cb(msg->to_string());
            });
            cli.subscribe(topic, qos)->wait();
            return true;
        } catch (...) {
            return false;
        }
    }

    bool Publish(const std::string& topic, const std::string& payload, int qos) {
        if (!is_connected) return false;
        try {
            mqtt::message_ptr pubmsg = mqtt::make_message(topic, payload);
            pubmsg->set_qos(qos);
            cli.publish(pubmsg)->wait();
            return true;
        } catch (...) {
            return false;
        }
    }

private:
    mqtt::async_client cli;
    std::atomic<bool> is_connected;
};

// ========== DeviceShifu APIs ==========

class DeviceShifu {
public:
    DeviceShifu()
    : stop_flag(false), phase(EdgeDevicePhase::Pending),
      edge_name(getenv_or_throw("EDGEDEVICE_NAME")),
      edge_ns(getenv_or_throw("EDGEDEVICE_NAMESPACE")),
      mqtt_broker(getenv_or_throw("MQTT_BROKER_ADDRESS")),
      mqtt_client(nullptr),
      config() {}

    ~DeviceShifu() {
        Stop();
    }

    void Start() {
        std::thread([this] { this->MonitorLoop(); }).detach();
    }

    void Stop() {
        stop_flag = true;
        if (mqtt_client) mqtt_client->Disconnect();
    }

    // ========== API Implementations ==========

    // Subscribe APIs
    void Subscribe_Odom(std::function<void(const std::string&)> cb) {
        SubscribeToTopic("device/sensors/odom", "SUBSCRIBE", cb);
    }
    void Subscribe_Imu(std::function<void(const std::string&)> cb) {
        SubscribeToTopic("device/sensors/imu", "SUBSCRIBE", cb);
    }
    void Subscribe_Laser(std::function<void(const std::string&)> cb) {
        SubscribeToTopic("device/sensors/laser", "SUBSCRIBE", cb);
    }

    // Publish APIs
    bool Publish_TTS(const std::string& json_payload) {
        return PublishToTopic("device/commands/tts", "PUBLISH", json_payload);
    }
    bool Publish_CmdVel(const std::string& json_payload) {
        return PublishToTopic("device/commands/cmd_vel", "PUBLISH", json_payload);
    }

    // ========== Status Management ==========

    EdgeDevicePhase GetPhase() const {
        return phase;
    }

private:
    std::atomic<bool> stop_flag;
    std::atomic<EdgeDevicePhase> phase;
    std::string edge_name;
    std::string edge_ns;
    std::string mqtt_broker;
    std::unique_ptr<MQTTClientWrapper> mqtt_client;
    InstructionConfig config;
    K8sClient k8s_client;
    std::mutex mtx;
    std::condition_variable cv;

    static std::string getenv_or_throw(const char* key) {
        const char* val = std::getenv(key);
        if (!val) {
            std::cerr << "Missing required environment variable: " << key << std::endl;
            std::exit(1);
        }
        return std::string(val);
    }

    void MonitorLoop() {
        // Load config
        if (!config.Load("/etc/edgedevice/config/instructions")) {
            std::cerr << "Failed to load instruction config." << std::endl;
        }
        // Get device address (not used for MQTT connection, but required by spec)
        std::string device_address = k8s_client.GetEdgeDeviceAddress(edge_ns, edge_name);
        // MQTT client
        mqtt_client = std::make_unique<MQTTClientWrapper>(mqtt_broker, edge_name + "-shifu");
        while (!stop_flag) {
            bool connected = mqtt_client->Connect();
            EdgeDevicePhase new_phase = connected ? EdgeDevicePhase::Running : EdgeDevicePhase::Failed;
            {
                std::unique_lock<std::mutex> lock(mtx);
                if (phase != new_phase) {
                    phase = new_phase;
                    k8s_client.PatchEdgeDevicePhase(edge_ns, edge_name, phase);
                }
            }
            int sleep_sec = connected ? 30 : 5;
            for (int i = 0; i < sleep_sec && !stop_flag; ++i) std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        k8s_client.PatchEdgeDevicePhase(edge_ns, edge_name, EdgeDevicePhase::Unknown);
    }

    void SubscribeToTopic(const std::string& topic, const std::string& api, std::function<void(const std::string&)> cb) {
        const APISettings* settings = config.GetSettings(api);
        int qos = 1;
        if (settings) {
            auto it = settings->protocolPropertyList.find("qos");
            if (it != settings->protocolPropertyList.end()) {
                try { qos = std::stoi(it->second); } catch (...) {}
            }
        }
        if (mqtt_client && mqtt_client->IsConnected()) {
            mqtt_client->Subscribe(topic, qos, cb);
        }
    }

    bool PublishToTopic(const std::string& topic, const std::string& api, const std::string& payload) {
        const APISettings* settings = config.GetSettings(api);
        int qos = 1;
        if (settings) {
            auto it = settings->protocolPropertyList.find("qos");
            if (it != settings->protocolPropertyList.end()) {
                try { qos = std::stoi(it->second); } catch (...) {}
            }
        }
        if (mqtt_client && mqtt_client->IsConnected()) {
            return mqtt_client->Publish(topic, payload, qos);
        }
        return false;
    }
};

// ========== Main Entrypoint ==========

static DeviceShifu* global_shifu = nullptr;

void signal_handler(int signum) {
    if (global_shifu) {
        global_shifu->Stop();
    }
    std::_Exit(0);
}

int main() {
    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    DeviceShifu shifu;
    global_shifu = &shifu;
    shifu.Start();

    // Example usage loop (replace with real application logic)
    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(60));
    }
    return 0;
}
```
